third_party/eigen3/complex.patch

```diff
@@ -0,0 +1,49 @@
+diff -Naur LOST.DIR/Complex.h install/Complex.h
+--- LOST.DIR/Complex.h	2023-07-18 13:33:18.000000000 +0000
++++ install/Complex.h	2023-07-18 13:31:02.000000000 +0000
+@@ -42,6 +42,7 @@
+ #if !(EIGEN_COMP_ICC && defined(_USE_COMPLEX_SPECIALIZATION_))
+ 
+ // Import Eigen's internal operator specializations.
++#if !(defined PLATFORM_WINDOWS && defined TENSORFLOW_CORE_KERNELS_SEGMENT_REDUCTION_OPS_GPU_CU_H_)
```

Comment:
If there's an issue here, we should probably fix it within Eigen, not with a TF-reduction-specific patch. 

What is the specific error addressed here?  We do test these complex operators on device with reduction kernels in Eigen compiled with MSVC.

Reply:
This patch is to resolve the following operator overloading errors:

> error: more than one operator "+=" "-=" "*=" "/=" matches these operands:
> function "Eigen::complex_operator_detail::operator += -= *= /= (std::complex<float> &, const std::complex<float> &)"
> E:\_bazel_tensorflow\4zvk5ci6\execroot\org_tensorflow\external\eigen_archive\Eigen\src/Core/arch/GPU/Complex.h(247): here
> function "std::complex<float>::operator += -= *= /= (const std::complex<float> &)"
> C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/include\complex(1055): here
> operand types are: tsl::complex64 *= const tsl::complex64

In Eigen GPU Complex.h, they said the followings:

> Compiling with MSVC and using +=,-=,*=,/=(std::complex<Scalar>) will lead to duplicate definition errors, since these are already specialized in Visual Studio's <complex> header (contrary to the standard). This is preferable to removing such definitions, which will lead to silent kernel failures.

`AlignedVector` is a template class having no data type. I don't know whether I can specify operators of complex via `using` inside the class. Also, I don't know if using std::complex operators is a correct choice.

Reply:
> Compiling with MSVC and using +=,-=,*=,/=(std::complex) will lead to duplicate definition errors, since these are already specialized in Visual Studio's header (contrary to the standard). This is preferable to removing such definitions, which will lead to silent kernel failures.

Ah, yes, I remember now.  The issue is if you try to use `std::complex` operators on device, it was compiling successfully but would crash at runtime with MSVC.  So we figured it was better to have the compilation error here.  Maybe we can restrict the overloads to only the device-compile phase? or does that lead to the same duplicate definition error?

Otherwise, an alternative is to replace
```
Before:  x += y;
After:   x = x + y;
```
to explicitly avoid the issue on MSVC.  I believe we've made this change in other places to work around the issue before.

Failing that, we can add a flag to disable these operators in upstream Eigen, but it will likely lead to runtime errors.

Reply:
I have replaced the operators in https://github.com/tensorflow/tensorflow/pull/61339. Just review there.